lib/src/main/java/com/tbruyelle/rxpermissions/ShadowActivity.java

```diff
@@ -23,12 +28,32 @@ protected void onNewIntent(Intent intent) {
     }
 
     private void handleIntent(Intent intent) {
-        String[] permissions = intent.getStringArrayExtra("permissions");
-        requestPermissions(permissions, 42);
+        List<String> perms = Arrays.asList(intent.getStringArrayExtra("permissions"));
+        List<String> yetGrantedPerms = Collections.emptyList();
+
+        // Figure out what permissions are not granted.
+        for (String perm : perms) {
+            int check = ContextCompat.checkSelfPermission(this,
+                                                          perm);
+
+            if (PackageManager.PERMISSION_GRANTED != check) {
+                yetGrantedPerms.add(perm);
+            }
+        }
+
+        // Request permissions if any.
+        int permNum = yetGrantedPerms.size();
+        if (permNum > 0) {
+            ActivityCompat.requestPermissions(this,
+                                              yetGrantedPerms.toArray(new String[permNum]),
+                                              42);
+        }
```

Comment:
We perform similar checks in RxPermissions class, before starting ShadowActivity.
I believe that we don't need to recheck permissions state here.